hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/execution/bulkinsert/RDDCustomColumnsSortPartitioner.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.execution.bulkinsert;
+
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.hudi.common.model.HoodieRecord;
+import org.apache.hudi.common.model.HoodieRecordPayload;
+import org.apache.hudi.table.BulkInsertPartitioner;
+import org.apache.spark.api.java.JavaRDD;
+
+/**
+ * A partitioner that does sorting based on specified column values for each RDD partition.
+ *
+ * @param <T> HoodieRecordPayload type
+ */
+public class RDDCustomColumnsSortPartitioner<T extends HoodieRecordPayload>
+    implements BulkInsertPartitioner<JavaRDD<HoodieRecord<T>>> {
+
+  private final String[] sortColumnNames;
+  private final String schemaString;
+
+  public RDDCustomColumnsSortPartitioner(String[] columnNames, Schema schema) {
+    this.sortColumnNames = columnNames;
+    //TODO Schema is not serializable. So convert to String here. Figure out how to improve this
+    this.schemaString = schema.toString();
+  }
+
+  @Override
+  public JavaRDD<HoodieRecord<T>> repartitionRecords(JavaRDD<HoodieRecord<T>> records,
+                                                     int outputSparkPartitions) {
+    final String[] sortColumns = this.sortColumnNames;
+    final String schemaStr = this.schemaString;
+    return records.sortBy(record -> {
```

Comment:
I have a suggestion to change the way we are generating the sort cols. Instead of performing this operation during write where the following code will spend cycles to convert data to generic record again, can we do this during read time when the HoodieRecord is being constructed ? I think you are using HoodieFileSliceReader and HoodieMergedRecordScanner to do this, does it make sense to add a sortKey to HoodieRecord and when you are constructing the hoodie record where you anyways have a handle to the generic record, you can just set that value in the hoodie record ? 
The downside of this approach is that the RDD bloats up so the shuffle will be larger so the question is whether that's worse or converting to generic record is worse from a CPU perspective

Reply:
I like this suggestion. This 'sortColumn' is probably needed only for this 'BulkInsert' based clustering strategy.  So, just want to make sure its ok to add new fields in HoodieRecord.  If you think thats acceptable, I'll make the change. Let me know.

Reply:
@satishkotha Do we know the overhead of converting schemaString to schema and generic record ?

Reply:
@n3nash I dont have numbers for this. I'll add metrics and calculate this. If you have any details from past examples, let me know